pkg/kubelet/network/kubenet/kubenet_linux.go

```diff
@@ -456,3 +487,315 @@ func (plugin *kubenetNetworkPlugin) getNsenterPath() (string, error) {
+	// this so that it is easier to flush and change, for example if the mark
+	// value should ever change.
+	writeLine(natRules, []string{
+		"-A", string(kubenetPostroutingChain),
+		"-m", "comment", "--comment", `"kubenet hostport traffic requiring SNAT"`,
+		"-m", "mark", "--mark", masqueradeMark,
+		"-j", "MASQUERADE",
+	}...)
+
+	// Install the kubenet-specific masquerade mark rule. We use a whole chain for
+	// this so that it is easier to flush and change, for example if the mark
+	// value should ever change.
+	writeLine(natRules, []string{
+		"-A", string(kubenetMarkChain),
+		"-j", "MARK", "--set-xmark", masqueradeMark,
+	}...)
+
+	// Accumulate NAT chains to keep.
+	activeNATChains := map[utiliptables.Chain]bool{} // use a map as a set
+
+	for containerPort, target := range containerPortMap {
+		protocol := strings.ToLower(string(containerPort.Protocol))
+		hostportChain := hostportChainName(containerPort, target.podFullName)
+		if chain, ok := existingNATChains[hostportChain]; ok {
+			writeLine(natChains, chain)
+		} else {
+			writeLine(natChains, utiliptables.MakeChainLine(hostportChain))
+		}
+
+		activeNATChains[hostportChain] = true
+
+		// Capture the hostport
+		args := []string{
+			"-A", string(kubenetHostportChain),
+			"-m", "comment", "--comment", fmt.Sprintf(`"%s hostport %d"`, target.podFullName, containerPort.HostPort),
+			"-m", protocol, "-p", protocol,
+			"--dport", fmt.Sprintf("%d", containerPort.HostPort),
+		}
+
+		writeLine(natRules, append(args, "-j", string(kubenetMarkChain))...)
+		writeLine(natRules, append(args, "-j", string(hostportChain))...)
+
+		// Create hostport chain to DNAT traffic to final destination
+		args = []string{
+			"-A", string(hostportChain),
+			"-m", "comment", "--comment", fmt.Sprintf(`"%s hostport %d"`, target.podFullName, containerPort.HostPort),
+			"-m", protocol, "-p", protocol,
+			"-j", "DNAT", fmt.Sprintf("--to-destination=%s:%d", target.podIP, containerPort.ContainerPort),
+		}
+
+		writeLine(natRules, args...)
+	}
+
+	// Delete chains no longer in use.
+	for chain := range existingNATChains {
+		if !activeNATChains[chain] {
+			chainString := string(chain)
+			if !strings.HasPrefix(chainString, kubenetHostportChainPrefix) {
+				// Ignore chains that aren't ours.
+				continue
+			}
+			// We must (as per iptables) write a chain-line for it, which has
+			// the nice effect of flushing the chain.  Then we can remove the
+			// chain.
+			writeLine(natChains, existingNATChains[chain])
+			writeLine(natRules, "-X", chainString)
+		}
+	}
+	writeLine(natRules, "COMMIT")
+
+	natLines := append(natChains.Bytes(), natRules.Bytes()...)
+	glog.V(3).Infof("Restoring iptables rules: %s", natLines)
+	err = plugin.iptables.RestoreAll(natLines, utiliptables.NoFlushTables, utiliptables.RestoreCounters)
+	if err != nil {
+		glog.Errorf("Failed to execute iptables-restore: %v", err)
+		return
+	}
+
+	plugin.syncHostportMap(containerPortMap)
```

Comment:
If you do this AFTER the iptables you will potentially install an iptables rule for a port that is in-use.  I think you have to open the local ports BEFORE you do the iptables-restore, like kube-proxy.